language/sygus.tex

```diff
@@ -1068,29 +1083,52 @@ \subsection{Asserting Synthesis Constraints and Assumptions}
 where $(x_1, \ldots, x_n)$ and $(y_1, \ldots, y_n)$
 are tuples of variables of sort $(\sigma_1, \ldots, \sigma_n)$ and
 $\varphi_{pre}$, $\varphi_{trans}$ and $\varphi_{post}$ are formulas.
-Given that this command is well-formed, given the above definitions,
+
+When this command is well-formed, given the above definitions,
 this command is syntax sugar for:
-\[
-\begin{array}{l}
-   \paren{\vardeclkwd\mbox{ }v_1\mbox{ }\sigma_1}\\
-   \paren{\vardeclkwd\mbox{ }v'_1\mbox{ }\sigma_1}\\
-   \ldots\\
-   \paren{\vardeclkwd\mbox{ }v_n\mbox{ }\sigma_n}\\
-   \paren{\vardeclkwd\mbox{ }v'_n\mbox{ }\sigma_n}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S_{pre}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{\texttt{and}\mbox{ }\paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{trans}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}\mbox{ }
-   \paren{S\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{post}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-\end{array}
-\]
+\begin{lstlisting}[language=SyGuS, numbers=none, basicstyle=\ttfamily\small]
+($\vardeclkwd$ $v_1$ $\sigma_1$)
+($\vardeclkwd$ $v'_1$ $\sigma_1$)
+$\ldots$
+($\vardeclkwd$ $v_n$ $\sigma_n$)
+($\vardeclkwd$ $v'_n$ $\sigma_n$)
+($\constraintkwd$ (=> ($S_{pre}$ $v_1$ $\ldots$ $v_n$) ($S$ $v_1$ $\ldots$ $v_n$)))
+($\constraintkwd$ (=> (and ($S$ $v_1$ $\ldots$ $v_n$) ($S_{trans}$ $v_1$ $\ldots$ $v_n$ $v'_1$ $\ldots$ $v'_n$)) ($S$ $v'_1$ $\ldots$ $v'_n$)))
+($\constraintkwd$ (=> ($S$ $v_1$ $\ldots$ $v_n$) ($S_{post}$ $v_1$ $\ldots$ $v_n$)))
+\end{lstlisting}
 where $v_1, v'_1, \ldots, v_n, v'_n$ are fresh symbols.
 
+\item \sygusinline|($\constraintchckwd$ (($x_1$ $\sigma_1$) $\ldots$ ($x_m$ $\sigma_m$)) $t_{tail}$ $t_{head}$)|
+
+This command adds a constrained Horn clause (CHC) constraint to the current state.
+Concretely, this command adds $m$ fresh variables $v_1, \ldots, v_m$ to the list of universal variables,
+and a constraint $((\lambda x_1,\ldots,x_m \ldotp t_{body})\ v_1\ \ldots\ v_m) \implies ((\lambda x_1,\ldots,x_m \ldotp t_{head})\ v_1\ \ldots\ v_m)$
+to the current state.
+This is a generalization of the $\constraintinvkwd$ command,
+that allows for specifying problems with multiple, possibly interdependent,
+invariants to be synthesized.
+
+A constraint of this form is well-formed iff:
+\begin{enumerate}
+  \item the \emph{head} of the CHC (i.e., $t_{head}$) is a well-sorted term of sort $\sbool$,
+  \item the \emph{body} of the CHC (i.e., $t_{tail}$) is either \sygusinline|$\truekwd$|,
```

Comment:
This is actually a bit misleading if `n=1` since we still support single literals as tails right? Writing e.g. `(and (= x y))` is not proper syntax.

I think this could be omitted, just say that the body is also a well-sorted term of sort Bool, then perhaps say informally that we expect it to be "and", referencing the logic section.

Reply:
How about I change the `true` part to a well-sorted term of sort Bool, so something like:

> \item the \emph{body} of the CHC (i.e., $t_{tail}$) is either \sygusinline|$t_1$|,
> or is a well-sorted conjunctive formula \sygusinline|(and $t_1$ $\ldots$ $t_n$)|,
> where $t_1$, $\ldots$, $t_n$ are well-sorted terms of sort $\sbool$

Reply:
Then it's a bit redundant since t_1 could be a conjunction itself?

The thing I'd like to avoid is hardcoding a notion of Boolean connective here.

Reply:
> Then it's a bit redundant since t_1 could be a conjunction itself?

Yes, this would be redundant in that case (but not wrong).

I think I am unclear on the scope of this "well-formedness" paragraph vs the "logic" paragraph later. I wanted this paragraph to keep the syntactic constraints here, and have the semantic constraints (w.r.t. predicates to be synthesized) in the "logic" paragraph.

To specify the syntactic constraint here, I want to say that `tbody` must syntactically be a conjunction.

Reply:
@ajreynol Please take another look. I think I understand your concerns now, and I have tried to update the "well-formedness" para and the "logic" para to try and address them.

Please read both parts, because they have changed quite a bit. Thanks!